File: server/src/http.go
package main

import (
	"net/http"
	"os"
	"path"
	"strconv"
	"text/template"
	"time"

	"github.com/didip/tollbooth"
	"github.com/didip/tollbooth_gin"
	sentrygin "github.com/getsentry/sentry-go/gin"
	"github.com/gin-contrib/gzip"
	gsessions "github.com/gin-contrib/sessions"
	"github.com/gin-contrib/sessions/cookie"
	"github.com/gin-gonic/gin"
)

type TemplateData struct {
	// Shared
	WebsiteName string
	Title       string // Used to populate the "<title>" tag
	Domain      string // Used to validate that the user is going to the correct URL
	Version     int
	Compiling   bool // True if we are currently recompiling the TypeScript client
	WebpackPort int

	// Profile
	Name       string
	NamesTitle string

	// History
	History      []*GameHistory
	SpecificSeed bool
	Tags         map[int][]string

	// Scores
	DateJoined                 string
	NumGames                   int
	TimePlayed                 string
	NumGamesSpeedrun           int
	TimePlayedSpeedrun         string
	NumMaxScores               int
	TotalMaxScores             int
	PercentageMaxScores        string
	RequestedNumPlayers        int      // Used on the "Missing Scores" page
	NumMaxScoresPerType        []int    // Used on the "Missing Scores" page
	PercentageMaxScoresPerType []string // Used on the "Missing Scores" page
	SharedMissingScores        bool     // Used on the "Missing Scores" page
	VariantStats               []*UserVariantStats

	// Stats
	NumVariants int
	Variants    []*VariantStatsData

	// Variants
	BestScores    []int
	MaxScoreRate  string
	MaxScore      int
	AverageScore  string
	NumStrikeouts int
	StrikeoutRate string
	RecentGames   []*GameHistory
}

const (
	// The name supplied to the Gin session middleware can be any arbitrary string
	HTTPSessionName    = "hanabi.sid"
	HTTPSessionTimeout = 60 * 60 * 24 * 365 // 1 year in seconds
	HTTPReadTimeout    = 5 * time.Second
	HTTPWriteTimeout   = 10 * time.Second
)

var (
	domain       string
	useTLS       bool
	GATrackingID string
	webpackPort  int

	// HTTPClientWithTimeout is used for sending web requests to external sites,
	// which is used in various middleware
	// We don't want to use the default http.Client because it has no default timeout set
	HTTPClientWithTimeout = &http.Client{
		Timeout: HTTPWriteTimeout,
	}
)

func httpInit() {
	// Read some configuration values from environment variables
	// (they were loaded from the ".env" file in "main.go")
	domain = os.Getenv("DOMAIN")
	if len(domain) == 0 {
		logger.Info("The \"DOMAIN\" environment variable is blank; aborting HTTP initialization.")
		return
	}
	sessionSecret := os.Getenv("SESSION_SECRET")
	if len(sessionSecret) == 0 {
		logger.Info("The \"SESSION_SECRET\" environment variable is blank; " +
			"aborting HTTP initialization.")
		return
	}
	portString := os.Getenv("PORT")
	var port int
	if len(portString) == 0 {
		port = 80
	} else {
		if v, err := strconv.Atoi(portString); err != nil {
			logger.Fatal("Failed to convert the \"PORT\" environment variable to a number.")
			return
		} else {
			port = v
		}
	}
	tlsCertFile := os.Getenv("TLS_CERT_FILE")
	tlsKeyFile := os.Getenv("TLS_KEY_FILE")
	if len(tlsCertFile) != 0 && len(tlsKeyFile) != 0 {
		useTLS = true
		if port == 80 {
			port = 443
		}
	}
	GATrackingID = os.Getenv("GA_TRACKING_ID")
	webpackPortString := os.Getenv("WEBPACK_DEV_SERVER_PORT")
	if len(webpackPortString) == 0 {
		webpackPort = 8080
	} else {
		if v, err := strconv.Atoi(webpackPortString); err != nil {
			logger.Fatal("Failed to convert the \"WEBPACK_DEV_SERVER_PORT\" environment variable to a number.")
			return
		} else {
			webpackPort = v
		}
	}

	// Create a new Gin HTTP router
	gin.SetMode(gin.ReleaseMode)                       // Comment this out to debug HTTP stuff
	httpRouter := gin.Default()                        // Has the "Logger" and "Recovery" middleware attached
	httpRouter.Use(gzip.Gzip(gzip.DefaultCompression)) // Add GZip compression middleware

	// Attach rate-limiting middleware from Tollbooth
	// The limiter works per path request,
	// meaning that a user can only request one specific path every X seconds
	// Thus, this does not impact the ability of a user to download CSS and image files all at once
	// (However, we do not want to use the rate-limiter in development, since we might have multiple
	// tabs open that are automatically-refreshing with webpack-dev-server)
	if !isDev {
		limiter := tollbooth.NewLimiter(2, nil) // Limit each user to 2 requests per second
		limiter.SetMessage(http.StatusText(http.StatusTooManyRequests))
		limiterMiddleware := tollbooth_gin.LimitHandler(limiter)
		httpRouter.Use(limiterMiddleware)
	}

	// Create a session store
	httpSessionStore := cookie.NewStore([]byte(sessionSecret))
	options := gsessions.Options{
		Path:   "/",                // The cookie should apply to the entire domain
		MaxAge: HTTPSessionTimeout, // In seconds
	}
	if !isDev {
		// Bind the cookie to this specific domain for security purposes
		options.Domain = domain
		// Only send the cookie over HTTPS:
		// https://www.owasp.org/index.php/Testing_for_cookies_attributes_(OTG-SESS-002)
		options.Secure = useTLS
		// Mitigate XSS attacks:
		// https://www.owasp.org/index.php/HttpOnly
		options.HttpOnly = true
		// Mitigate CSRF attacks:
		// https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#SameSite_cookies
		options.SameSite = http.SameSiteStrictMode
	}
	httpSessionStore.Options(options)

	// Attach the sessions middleware
	httpRouter.Use(gsessions.Sessions(HTTPSessionName, httpSessionStore))

	// Initialize Google Analytics
	if len(GATrackingID) > 0 {
		httpRouter.Use(httpGoogleAnalytics) // Attach the Google Analytics middleware
	}

	// Attach the Sentry middleware
	if usingSentry {
		httpRouter.Use(sentrygin.New(sentrygin.Options{
			// https://github.com/getsentry/sentry-go/blob/master/gin/sentrygin.go
			Repanic: true, // Recommended as per the documentation
			Timeout: HTTPWriteTimeout,
		}))
		httpRouter.Use(sentryHTTPAttachMetadata)
	}

	// Path handlers (for cookies and logging in)
	httpRouter.POST("/login", httpLogin)
	httpRouter.GET("/logout", httpLogout)
	httpRouter.GET("/test-cookie", httpTestCookie)
	httpRouter.GET("/ws", httpWS)

	// Path handlers (for the main website)
	httpRouter.GET("/", httpMain)
	httpRouter.GET("/replay", httpMain)
	httpRouter.GET("/replay/:gameID", httpMain)
	httpRouter.GET("/replay/:gameID/:turn", httpMain)
	httpRouter.GET("/shared-replay", httpMain)
	httpRouter.GET("/shared-replay/:gameID", httpMain)
	httpRouter.GET("/shared-replay/:gameID/:turn", httpMain)
	httpRouter.GET("/create-table", httpMain)
	httpRouter.GET("/test", httpMain)
	httpRouter.GET("/test/:testNum", httpMain)

	// Path handlers (for development)
	// ("/dev" is the same as "/" but uses webpack-dev-server to serve JavaScript)
	httpRouter.GET("/dev", httpMain)
	httpRouter.GET("/dev/", httpMain)
	httpRouter.GET("/dev/replay", httpMain)
	httpRouter.GET("/dev/replay/:gameID", httpMain)
	httpRouter.GET("/dev/replay/:gameID/:turn", httpMain)
	httpRouter.GET("/dev/shared-replay", httpMain)
	httpRouter.GET("/dev/shared-replay/:gameID", httpMain)
	httpRouter.GET("/dev/shared-replay/:gameID/:turn", httpMain)
	httpRouter.GET("/dev/create-table", httpMain)
	httpRouter.GET("/dev/test", httpMain)
	httpRouter.GET("/dev/test/:testNum", httpMain)

	// Path handlers for other URLs
	httpRouter.GET("/scores", httpScores)
	httpRouter.GET("/scores/:player1", httpScores)
	httpRouter.GET("/profile", httpScores) // "/profile" is an alias for "/scores"
	httpRouter.GET("/profile/:player1", httpScores)
	httpRouter.GET("/history", httpHistory)
	httpRouter.GET("/history/:player1", httpHistory)
	httpRouter.GET("/history/:player1/:player2", httpHistory)
	httpRouter.GET("/history/:player1/:player2/:player3", httpHistory)
	httpRouter.GET("/history/:player1/:player2/:player3/:player4", httpHistory)
	httpRouter.GET("/history/:player1/:player2/:player3/:player4/:player5", httpHistory)
	httpRouter.GET("/history/:player1/:player2/:player3/:player4/:player5/:player6", httpHistory)
	httpRouter.GET("/missing-scores", httpMissingScores)
	httpRouter.GET("/missing-scores/:player1", httpMissingScores)
	httpRouter.GET("/missing-scores/:player1/:numPlayers", httpMissingScores)
	httpRouter.GET("/shared-missing-scores", httpSharedMissingScores)
	httpRouter.GET("/shared-missing-scores/:player1", httpSharedMissingScores)
	httpRouter.GET("/shared-missing-scores/:player1/:player2", httpSharedMissingScores)
	httpRouter.GET("/shared-missing-scores/:player1/:player2/:player3", httpSharedMissingScores)
	httpRouter.GET("/shared-missing-scores/:player1/:player2/:player3/:player4", httpSharedMissingScores)
	httpRouter.GET("/shared-missing-scores/:player1/:player2/:player3/:player4/:player5", httpSharedMissingScores)
	httpRouter.GET("/shared-missing-scores/:player1/:player2/:player3/:player4/:player5/:player6", httpSharedMissingScores)
	httpRouter.GET("/tags", httpTags)
	httpRouter.GET("/tags/:player1", httpTags)
	httpRouter.GET("/seed", httpSeed)
	httpRouter.GET("/seed/:seed", httpSeed) // Display all games played on a given seed
	httpRouter.GET("/stats", httpStats)
	httpRouter.GET("/variant", httpVariant)
	httpRouter.GET("/variant/:id", httpVariant)
	httpRouter.GET("/tag", httpTag)
	httpRouter.GET("/tag/:tag", httpTag)
	httpRouter.GET("/videos", httpVideos)
	httpRouter.GET("/password-reset", httpPasswordReset)
	httpRouter.POST("/password-reset", httpPasswordResetPost)

	// Path handlers for bots, developers, researchers, etc.
	httpRouter.GET("/export", httpExport)
	httpRouter.GET("/export/:game", httpExport)

	// Other
	httpRouter.Static("/public", path.Join(projectPath, "public"))
	httpRouter.StaticFile("/favicon.ico", path.Join(projectPath, "public", "img", "favicon.ico"))

	if useTLS {
		// Create the LetsEncrypt directory structure
		// (CertBot will look for data in "/.well-known/acme-challenge/####")
		letsEncryptPath := path.Join(projectPath, "letsencrypt")
		wellKnownPath := path.Join(letsEncryptPath, ".well-known")
		acmeChallengePath := path.Join(wellKnownPath, "acme-challenge")
		if _, err := os.Stat(letsEncryptPath); os.IsNotExist(err) {
			if err := os.MkdirAll(acmeChallengePath, 0755); err != nil {
				logger.Fatal("Failed to create the \""+acmeChallengePath+"\" directory:", err)
			}
		}
		if _, err := os.Stat(wellKnownPath); os.IsNotExist(err) {
			if err := os.MkdirAll(acmeChallengePath, 0755); err != nil {
				logger.Fatal("Failed to create the \""+acmeChallengePath+"\" directory:", err)
			}
		}
		if _, err := os.Stat(acmeChallengePath); os.IsNotExist(err) {
			if err := os.MkdirAll(acmeChallengePath, 0755); err != nil {
				logger.Fatal("Failed to create the \""+acmeChallengePath+"\" directory:", err)
			}
		}

		// We want all HTTP requests to be redirected to HTTPS
		// (but make an exception for Let's Encrypt)
		// The Gin router is using the default serve mux,
		// so we need to create a new fresh one for the HTTP handler
		HTTPServeMux := http.NewServeMux()
		HTTPServeMux.Handle(
			"/.well-known/acme-challenge/",
			http.FileServer(http.FileSystem(http.Dir(letsEncryptPath))),
		)
		HTTPServeMux.Handle(
			"/",
			http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
				http.Redirect(
					w,
					req,
					"https://"+req.Host+req.URL.String(),
					http.StatusMovedPermanently,
				)
			}),
		)

		// ListenAndServe is blocking, so we need to start listening in a new goroutine
		go func() {
			// We need to create a new http.Server because the default one has no timeouts
			// https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/
			HTTPRedirectServerWithTimeout := &http.Server{
				Addr:         "0.0.0.0:80", // Listen on all IP addresses
				Handler:      HTTPServeMux,
				ReadTimeout:  HTTPReadTimeout,
				WriteTimeout: HTTPWriteTimeout,
			}
			if err := HTTPRedirectServerWithTimeout.ListenAndServe(); err != nil {
				logger.Fatal("ListenAndServe failed to start on port 80.")
				return
			}
			logger.Fatal("ListenAndServe ended for port 80.")
		}()
	}

	// Start listening and serving requests (which is blocking)
	// We need to create a new http.Server because the default one has no timeouts
	// https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/
	logger.Info("Listening on port " + strconv.Itoa(port) + ".")
	HTTPServerWithTimeout := &http.Server{
		Addr:         "0.0.0.0:" + strconv.Itoa(port), // Listen on all IP addresses
		Handler:      httpRouter,
		ReadTimeout:  HTTPReadTimeout,
		WriteTimeout: HTTPWriteTimeout,
	}
	if useTLS {
		if err := HTTPServerWithTimeout.ListenAndServeTLS(tlsCertFile, tlsKeyFile); err != nil {
			logger.Fatal("ListenAndServeTLS failed:", err)
			return
		}
		logger.Fatal("ListenAndServeTLS ended prematurely.")
	} else {
		if err := HTTPServerWithTimeout.ListenAndServe(); err != nil {
			logger.Fatal("ListenAndServe failed:", err)
			return
		}
		logger.Fatal("ListenAndServe ended prematurely.")
	}
}

// httpServeTemplate combines a standard HTML header with the body for a specific page
// (we want the same HTML header for all pages)
func httpServeTemplate(w http.ResponseWriter, data TemplateData, templateName ...string) {
	// Since we are using the GZip middleware, we have to specify the content type,
	// or else the page will be downloaded by the browser as "download.gz"
	w.Header().Set("Content-Type", "text/html; charset=utf-8")

	viewsPath := path.Join(projectPath, "server", "src", "views")
	layoutPath := path.Join(viewsPath, "layout.tmpl")
	logoPath := path.Join(viewsPath, "logo.tmpl")

	// Turn the slice of file names into a slice of full paths
	for i := 0; i < len(templateName); i++ {
		templateName[i] = path.Join(viewsPath, templateName[i]+".tmpl")
	}

	// Ensure that the layout file exists
	if _, err := os.Stat(layoutPath); os.IsNotExist(err) {
		logger.Error("The layout template does not exist.")
		http.Error(
			w,
			http.StatusText(http.StatusInternalServerError),
			http.StatusInternalServerError,
		)
		return
	}

	// Return a 404 if the template doesn't exist or it is a directory
	if info, err := os.Stat(layoutPath); os.IsNotExist(err) {
		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
		return
	} else if info.IsDir() {
		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
		return
	}

	// Append the main layout to our list of layouts
	templateName = append(templateName, layoutPath)

	// Append the nav bar logo to our list of layouts
	templateName = append(templateName, logoPath)

	// Create the template
	var tmpl *template.Template
	if v, err := template.New("template").Funcs(template.FuncMap{
		"formatDate": httpFormatDate,
	}).ParseFiles(templateName...); err != nil {
		logger.Error("Failed to create the template:", err.Error())
		http.Error(
			w,
			http.StatusText(http.StatusInternalServerError),
			http.StatusInternalServerError,
		)
		return
	} else {
		tmpl = v
	}

	// Add extra data that should be the same for every page request
	data.WebsiteName = WebsiteName

	// Execute the template and send it to the user
	if err := tmpl.ExecuteTemplate(w, "layout", data); err != nil {
		if isCommonHTTPError(err.Error()) {
			logger.Info("Ordinary error when executing the template: " + err.Error())
		} else {
			logger.Error("Failed to execute the template: " + err.Error())
		}
		http.Error(
			w,
			http.StatusText(http.StatusInternalServerError),
			http.StatusInternalServerError,
		)
	}
}

func httpFormatDate(date time.Time) string {
	return date.Format("2006-01-02 &mdash; 15:04:05 MST")
}

File: packages/server/src/http.ts
import fastifyFormBody from "@fastify/formbody";
import type { SecureSessionPluginOptions } from "@fastify/secure-session";
import { fastifySecureSession } from "@fastify/secure-session";
import { fastifyStatic } from "@fastify/static";
import { fastifyView } from "@fastify/view";
import fastifyWebSocket from "@fastify/websocket";
import { Eta } from "eta";
import type {
  FastifyInstance,
  RawReplyDefaultExpression,
  RawRequestDefaultExpression,
  RawServerDefault,
} from "fastify";
import { fastify } from "fastify";
import { fastifyFavicon } from "fastify-favicon";
import { StatusCodes } from "http-status-codes";
import fs from "node:fs";
import path from "node:path";
import type { Logger } from "pino";
import { isBannedIP } from "./bannedIPs";
import { REPO_ROOT } from "./constants";
import { IS_DEV, env } from "./env";
import { httpLogin } from "./http/httpLogin";
import { httpLogout } from "./http/httpLogout";
import { httpMain } from "./http/httpMain";
import { httpTestCookie } from "./http/httpTestCookie";
import { httpWS } from "./http/httpWS";
import { logger } from "./logger";

type FastifyInstanceWithLogger = FastifyInstance<
  RawServerDefault,
  RawRequestDefaultExpression,
  RawReplyDefaultExpression,
  Logger
>;

const COOKIE_NAME = "hanabi.sid";

const COOKIE_OPTIONS_BASE = {
  /** The cookie should apply to the entire domain. */
  path: "/",

  /** 1 year in seconds. */
  maxAge: 60 * 60 * 24 * 365,
} as const satisfies SecureSessionPluginOptions["cookie"];

const COOKIE_OPTIONS_PRODUCTION = {
  ...COOKIE_OPTIONS_BASE,

  /** Bind the cookie to this specific domain for security purposes. */
  domain: env.DOMAIN,

  /**
   * Only send the cookie over HTTPS:
   * https://www.owasp.org/index.php/Testing_for_cookies_attributes_(OTG-SESS-002)
   */
  secure: true,

  /**
   * Mitigate XSS attacks:
   * https://www.owasp.org/index.php/HttpOnly
   */
  httpOnly: true,

  /**
   * Mitigate CSRF attacks:
   * https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#SameSite_cookies
   */
  sameSite: "strict",
} as const satisfies SecureSessionPluginOptions["cookie"];

const COOKIE_OPTIONS = IS_DEV ? COOKIE_OPTIONS_BASE : COOKIE_OPTIONS_PRODUCTION;

/**
 * The Let's Encrypt certbot will request a token from this path:
 * https://letsencrypt.org/docs/challenge-types/
 */
const LETS_ENCRYPT_PATH_PREFIX = "/letsencrypt/.well-known/acme-challenge";

const LETS_ENCRYPT_PATH = path.join(REPO_ROOT, LETS_ENCRYPT_PATH_PREFIX);

const eta = new Eta({
  // Prefer the Golang template syntax, which is not the default.
  tags: ["{{", "}}"],

  // By default, Eta will allow undefined values to bubble through to the template. We want to
  // explicitly throw an error in this case. This solution was recommended by Ben Gubler in the
  // Discord server.
  functionHeader: `
    if (it !== undefined) {
      for (const [key, value] of Object.entries(it)) {
        if (typeof value !== "string" && typeof value !== "number" && typeof value !== "boolean") {
          throw new Error('The template value of "' + key + '" has a type of: ' + typeof value);
        }

        if (typeof value === "string" && value.trim() === "") {
          throw new Error('The template value of "' + key + '" is empty.');
        }
      }
    }
  `,
});

export async function httpInit(): Promise<void> {
  const useTLS = env.TLS_CERT_FILE !== "" && env.TLS_KEY_FILE !== "";
  const defaultPort = useTLS ? 443 : 80;
  const port = env.PORT === 0 ? defaultPort : env.PORT;

  const https = useTLS
    ? {
        ca: env.TLS_CERT_FILE,
        key: env.TLS_KEY_FILE,
      }
    : null; // eslint-disable-line unicorn/no-null

  // Initialize the HTTP server using the Fastify library:
  // https://fastify.dev/
  const httpServer = fastify({
    logger,
    https,

    // The "incoming request" and "request completed" logs are too noisy when developing.
    disableRequestLogging: IS_DEV,
  });

  await registerFastifyPlugins(httpServer, useTLS);

  httpServer.setNotFoundHandler(async (_request, reply) => {
    await reply
      .code(StatusCodes.NOT_FOUND)
      .type("text/html")
      .send("404 not found");
  });

  httpServer.setErrorHandler(async (error, _request, reply) => {
    // Use "||" to handle undefined and an empty string at the same time.
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions, @typescript-eslint/prefer-nullish-coalescing
    const text = error.stack || error.message || "unknown error";

    await reply
      .code(StatusCodes.INTERNAL_SERVER_ERROR)
      .type("text/html")
      .send(`<pre>${text}</pre>`);
  });

  httpServer.addHook("preHandler", async (request, reply) => {
    const { ip } = request;

    if (isBannedIP(ip)) {
      logger.info(
        `IP "${request.ip}" tried to send an HTTP request, but they are banned.`,
      );
      return reply
        .code(StatusCodes.UNAUTHORIZED)
        .send(
          "Your IP address has been banned. Please contact an administrator if you think this is a mistake.",
        );
    }
  });

  registerPathHandlers(httpServer);

  await httpServer.listen({
    port,
  });
}

/** Plugins are registered in alphabetical order. */
async function registerFastifyPlugins(
  httpServer: FastifyInstanceWithLogger,
  useTLS: boolean,
) {
  // `fastify-favicon` - Needed for the favicon:
  // https://github.com/smartiniOnGitHub/fastify-favicon
  await httpServer.register(fastifyFavicon, {
    // The plugin appends "favicon.ico" to the provided path.
    path: path.join(REPO_ROOT, "public", "img"),
  });

  // `@fastify/formbody` - Needed for "application/x-www-form-urlencoded" POST data:
  // https://github.com/fastify/fastify-formbody
  await httpServer.register(fastifyFormBody);

  // `@fastify/secure-session` - Needed for session management:
  // https://github.com/fastify/fastify-secure-session/tree/master
  // (The `@fastify/session` plugin is used for server-side data storage.)
  await httpServer.register(fastifySecureSession, {
    cookieName: COOKIE_NAME,
    secret: env.SESSION_SECRET,
    salt: "hanabiSalt123456", // Must be 16 characters long.
    cookie: COOKIE_OPTIONS,
  });

  // `@fastify/websocket` - Needed for WebSockets.
  await httpServer.register(fastifyWebSocket, {
    options: {
      maxPayload: 123,
    },
  });

  // `@fastify/static` - Needed to serve static files:
  // https://github.com/fastify/fastify-static
  await httpServer.register(fastifyStatic, {
    root: path.join(REPO_ROOT, "public"),
    prefix: "/public",
  });

  // Handle renewing HTTPS certificates through `certbot`:
  // https://letsencrypt.org/docs/challenge-types/
  if (useTLS) {
    fs.mkdirSync(LETS_ENCRYPT_PATH, {
      recursive: true,
    });

    await httpServer.register(fastifyStatic, {
      root: LETS_ENCRYPT_PATH,
      prefix: LETS_ENCRYPT_PATH_PREFIX,
    });
  }

  // `@fastify/view` - Needed for the template library:
  // https://github.com/fastify/point-of-view
  await httpServer.register(fastifyView, {
    engine: {
      eta,
    },
    templates: path.join(__dirname, "templates"),
  });
}

function registerPathHandlers(httpServer: FastifyInstanceWithLogger) {
  // For cookies and logging in.
  httpServer.post("/login", httpLogin);
  httpServer.get("/logout", httpLogout);
  httpServer.get("/test-cookie", httpTestCookie);
  httpServer.get("/ws", { websocket: true }, httpWS);

  // For the main website.
  httpServer.get("/", httpMain);
  httpServer.get("/lobby", httpMain);
  httpServer.get("/pre-game", httpMain);
  httpServer.get("/pre-game/:tableID", httpMain);
  httpServer.get("/game", httpMain);
  httpServer.get("/game/:tableID", httpMain);
  httpServer.get("/game/:tableID/shadow/:seat", httpMain);
  httpServer.get("/replay", httpMain);
  httpServer.get("/replay/:databaseID", httpMain);
  httpServer.get("/shared-replay", httpMain);
  httpServer.get("/shared-replay/:databaseID", httpMain);
  httpServer.get("/replay-json/:string", httpMain);
  httpServer.get("/shared-replay-json/:string", httpMain);
  httpServer.get("/create-table", httpMain);

  /*

	// Path handlers for other URLs
	fastify.get("/scores", httpScores)
	fastify.get("/scores/:player1", httpScores)
	fastify.get("/profile", httpScores) // "/profile" is an alias for "/scores"
	fastify.get("/profile/:player1", httpScores)
	fastify.get("/history", httpHistory)
	fastify.get("/history/:player1", httpHistory)
	fastify.get("/history/:player1/:player2", httpHistory)
	fastify.get("/history/:player1/:player2/:player3", httpHistory)
	fastify.get("/history/:player1/:player2/:player3/:player4", httpHistory)
	fastify.get("/history/:player1/:player2/:player3/:player4/:player5", httpHistory)
	fastify.get("/history/:player1/:player2/:player3/:player4/:player5/:player6", httpHistory)
	fastify.get("/missing-scores", httpMissingScores)
	fastify.get("/missing-scores/:player1", httpMissingScores)
	fastify.get("/missing-scores/:player1/:numPlayers", httpMissingScores)
	fastify.get("/shared-missing-scores/:numPlayers", httpSharedMissingScores)
	fastify.get("/shared-missing-scores/:numPlayers/:player1", httpSharedMissingScores)
	fastify.get("/shared-missing-scores/:numPlayers/:player1/:player2", httpSharedMissingScores)
	fastify.get("/shared-missing-scores/:numPlayers/:player1/:player2/:player3", httpSharedMissingScores)
	fastify.get("/shared-missing-scores/:numPlayers/:player1/:player2/:player3/:player4", httpSharedMissingScores)
	fastify.get("/shared-missing-scores/:numPlayers/:player1/:player2/:player3/:player4/:player5", httpSharedMissingScores)
	fastify.get("/shared-missing-scores/:numPlayers/:player1/:player2/:player3/:player4/:player5/:player6", httpSharedMissingScores)
	fastify.get("/tags", httpTags)
	fastify.get("/tags/:player1", httpTags)
	fastify.get("/seed", httpSeed)
	fastify.get("/seed/:seed", httpSeed) // Display all games played on a given seed
	fastify.get("/stats", httpStats)
	fastify.get("/variant", httpVariant)
	fastify.get("/variant/:id", httpVariant)
	fastify.get("/tag", httpTag)
	fastify.get("/tag/:tag", httpTag)
	fastify.get("/videos", httpVideos)
	fastify.get("/password-reset", httpPasswordReset)
	httpRouter.POST("/password-reset", httpPasswordResetPost)

	// API V1 routes
	apiSetRoutes(httpRouter)

	// Path handlers for bots, developers, researchers, etc.
	fastify.get("/export", httpExport)
	fastify.get("/export/:databaseID", httpExport)

  */
}
